torch/testing/_internal/common_utils.py

```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
 
         return tg
 
-    def assertEqual(self, x, y, prec=None, message='', allow_inf=False, exact_dtype=None):
+    # Some analysis of tolerance by logging tests from test_torch.py can be found
+    # in https://github.com/pytorch/pytorch/pull/32538.
+    # dtype name : (rtol, atol)
+    dtype_precisions = {
+        'float16': (0.001, 1e-5),
+        'bfloat16': (0.016, 1e-5),
+        'float32': (1.3e-6, 1e-5),
+        'float64': (1e-7, 1e-7),
+        'complex32': (1.3e-6, 1e-5),
+        'complex64': (1e-7, 1e-7),
+    }
+
+    # todo: implement numpy-like issubdtype
+    def is_integral(self, dtype):
+        dtypes = get_all_dtypes()
+        return dtype in dtypes and not dtype.is_floating_point
+
+    def get_default_tolerance(self, a, b=None):
+        if b is None:
+            dtype = torch.float
+            if isinstance(a, torch.Tensor):
+                dtype = a.dtype
+            elif isinstance(a, torch.dtype):
+                dtype = a
+            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
+                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
+                    dtype = torch.float
+                else:
+                    dtype = torch.from_numpy(a).dtype
+            if self.is_integral(dtype):
+                return (0, 0)
+            dtype = str(dtype).split('.')[-1]
+            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
+
+        a_tol = self.get_default_tolerance(a)
+        b_tol = self.get_default_tolerance(b)
+        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
+
+    def assertEqual(self, x, y, message='', **kwargs):
+        self.assertIsNone(kwargs.get('prec', None), 'prec is no longer supported. Use atol or rtol.')
+        rtol = kwargs.get('rtol', None)
+        atol = kwargs.get('atol', None)
+        allow_inf = kwargs.get('allow_inf', False)
+        exact_dtype = kwargs.get('exact_dtype', None)
```

Comment:
It would be good to do some sort of test to make sure someone doesn't send a `kwarg` which we ignore. But there's even a better solution now that we're Python 3 only: use keyword only arguments `(self, x, y, *, rtol=..., ....)`

Reply:
ah when I wrote this we still had python2 support. I think I might also prefer to land that update separately just to minimize conflicts in the meantime.